system-parachains/asset-hubs/asset-hub-kusama/src/xcm_config.rs

```diff
@@ -552,17 +557,63 @@ pub mod bridging {
 			}
 		}
 
-		/// Reserve locations filter for `xcm_executor::Config::IsReserve`.
-		/// Locations from which the runtime accepts reserved assets.
-		pub type IsTrustedBridgedReserveLocationForConcreteAsset =
-			matching::IsTrustedBridgedReserveLocationForConcreteAsset<
-				UniversalLocation,
-				(
-					// allow receive DOT from AssetHubPolkadot
-					xcm_builder::Case<DotFromAssetHubPolkadot>,
-					// and nothing else
-				),
-			>;
+		/// Allow any asset native to the Polkadot or Ethereum ecosystems if it comes from Polkadot
+		/// Asset Hub.
+		pub type PolkadotOrEthereumAssetFromAssetHubPolkadot = RemoteAssetFromLocation<
+			(StartsWith<PolkadotEcosystem>, StartsWith<EthereumEcosystem>),
+			AssetHubPolkadot,
+		>;
+
+		// TODO: get this from `assets_common v0.17.1` when SDK deps are upgraded
+		/// Accept an asset if it is native to `AssetsAllowedNetworks` and it is coming from
+		/// `OriginLocation`.
+		pub struct RemoteAssetFromLocation<AssetsAllowedNetworks, OriginLocation>(
+			sp_std::marker::PhantomData<(AssetsAllowedNetworks, OriginLocation)>,
+		);
+		impl<
+				L: TryInto<Location> + Clone,
+				AssetsAllowedNetworks: Contains<Location>,
+				OriginLocation: Get<Location>,
+			> ContainsPair<L, L> for RemoteAssetFromLocation<AssetsAllowedNetworks, OriginLocation>
+		{
+			fn contains(asset: &L, origin: &L) -> bool {
+				let latest_asset: Location = if let Ok(location) = asset.clone().try_into() {
+					location
+				} else {
+					return false;
+				};
+				let latest_origin: Location = if let Ok(location) = origin.clone().try_into() {
+					location
+				} else {
+					return false;
+				};
+				let expected_origin = OriginLocation::get();
+				// ensure `origin` is expected `OriginLocation`
+				if !expected_origin.eq(&latest_origin) {
+					log::trace!(
+						target: "xcm::contains",
+						"RemoteAssetFromLocation asset: {:?}, origin: {:?} is not from expected {:?}",
+						latest_asset, latest_origin, expected_origin,
+					);
+					return false;
+				} else {
+					log::trace!(
+						target: "xcm::contains",
+						"RemoteAssetFromLocation asset: {latest_asset:?}, origin: {latest_origin:?}",
+					);
+				}
+
+				// ensure `asset` is from remote consensus listed in `AssetsAllowedNetworks`
+				AssetsAllowedNetworks::contains(&latest_asset)
+			}
+		}
```

Comment:
> Note 1: Consider if it's possible to externalize this helper struct to a common crate (to avoid duplication and help to maintainability of this codebase).
> Note 2: Kudos for adding `trace` logs. This is ALWAYS appreciated.

Reply:
> Consider if it's possible to externalize this helper struct to a common crate (to avoid duplication and help to maintainability of this codebase).

yes, it's already done, we just need to get it from newer crate version when we upgrade deps, see [this comment just above your comment](https://github.com/acatangiu/runtimes/blob/829cb81a7b25de603457a14d05bc5232e054c410/system-parachains/asset-hubs/asset-hub-polkadot/src/xcm_config.rs#L634C3-L636C24)